packages/dev/core/src/Buffers/buffer.align.ts

```diff
@@ -0,0 +1,168 @@
+import type { Nullable } from "core/types";
+import type { DataBuffer } from "./dataBuffer";
+import { Buffer, VertexBuffer } from "./buffer";
+
+const isLittleEndian = (() => {
+    const array = new Uint8Array(4);
+    const view = new Uint32Array(array.buffer);
+
+    return !!((view[0] = 1) & array[0]);
+})();
+
+declare module "./buffer" {
+    export interface VertexBuffer {
+        /**
+         * Gets the effective byte stride, that is the byte stride of the buffer that is actually sent to the GPU.
+         * It could be different from VertexBuffer.byteStride if a new buffer must be created under the hood because of the forceVertexBufferStrideMultiple4Bytes engine flag.
+         */
+        effectiveByteStride: number;
+
+        /**
+         * Gets the effective byte offset, that is the byte offset of the buffer that is actually sent to the GPU.
+         * It could be different from VertexBuffer.byteOffset if a new buffer must be created under the hood because of the forceVertexBufferStrideMultiple4Bytes engine flag.
+         */
+        effectiveByteOffset: number;
+
+        /**
+         * Gets the effective buffer, that is the buffer that is actually sent to the GPU.
+         * It could be different from VertexBuffer.getBuffer() if a new buffer must be created under the hood because of the forceVertexBufferStrideMultiple4Bytes engine flag.
+         */
+        effectiveBuffer: Nullable<DataBuffer>;
+
+        /** @internal */
+        _alignBuffer(): void;
+
+        /** @internal */
+        _alignedBuffer?: Buffer;
+    }
+}
+
+Object.defineProperty(VertexBuffer.prototype, "effectiveByteStride", {
+    get: function (this: VertexBuffer) {

```

Comment:
why not add this directly to VertexBuffer? Will eventually be less code added

Reply:
In fact, we don't want this file to be in effect for the WebGL engine, because we don't need it (ANGLE is doing whatever needs to be done under the hood). That's why it is in a separate file, we don't want to make the `buffer.ts` file bigger for WebGL whereas the code is not needed.

Reply:
This way will not work for UMD, because the file will not be included. If included, it will be available for all UMD versions (so - also for WebGL), because the side effects will be populated.

We can think of different ways to solve this. For example,the engine is already a member of VertexBuffer t Then you can check for the isWebGPU flag in Buffer directly and decide whether or not to run one thing or the other. The variables can stay undefined.

I don't think a WebGPUBuffer will work, because of the way the class is being used all over.

Reply:
My previous version handled WebGL / WebGPU through a feature flag, so I could revert to this code, but it would make the `VertexBuffer` class bigger...

cc @sebavan to know what to do.

Reply:
I added back the feature flag so that it can work in WebGL in UMD mode (I don't want to test isWebGPU because we may need this code also for Native).

Let me know what you think?

Reply:
just to be sure - UMD version will still have those functions populated. is it enough to have the flag just on this function?

Reply:
Yes, if we don't run the `_alignBuffer` method (thanks to the flag), `_alignedBuffer` will remain `undefined` and everything will work as expected for WebGL.